include/task_ring_buffer.h

```diff
@@ -0,0 +1,87 @@
+#pragma once
+
+#include <array>
+#include <atomic>
+#include <memory>
+
+#include "task.h"
+#include "types.h"
+
+namespace celerity::detail {
+
+constexpr unsigned long task_ringbuffer_size = 1024;
+
+template <unsigned long N>
+class task_ring_buffer {
+  public:
+	size_t get_total_task_count() const { return next_active_tid.load(); }
+	size_t get_current_task_count() const { return next_active_tid.load() - number_of_deleted_tasks; }
+
+	bool has_task(task_id tid) const {
+		return tid >= number_of_deleted_tasks && tid < next_active_tid.load(); //
+	}
+
+	task* get_task(task_id tid) const {
+		assert(has_task(tid));
+		return data[tid % N].get();
+	}
+
+	task_id reserve_new_tid() {
+		wait_for_available_slot();
+		auto ret = next_task_id;
+		next_task_id++;
+		return ret;
+	}
+
+	// task_id must have been reserved previously
+	void emplace(task_id tid, std::unique_ptr<task> task) {
```

Comment:
Naming bikeshedding: emplace usually means it will be constructed in place. Maybe `put`?

Reply:
It was like this in the original code (of task manager) that's why I kept it that way. I agree in principle though and can also change it.